### ISSUE TITLE
test stops at first assertion error

### ISSUE THREAD

**Issue by DexterMagnific** [libcheck/check#208]
Dear developer,

Is it possible that the test continues until finished, even if some assertion inside it failed ?

The idea behind this is that with the current behavior, you must fix the current issue before having the chance to see is there are still other issues in your program.

Reporting as much violations as possible will reduce the back and forth between coding and testing. 

Thanks

**Comment by makr** [comment]
Just my 2cents: I would vote against your proposal. The right solution here is to separate the assertions into different test cases and only have multiple assertions in a test case that depend on each other.

**Comment by mikkoi** [comment]
In practice I agree with @makr . If an assert fails in a test with many stages, the later stages are not reliable.
In principle, however, during one stage there can be several variables that are not connected with each other and they all need to be checked. In a situation like that reporting all failures at one go would be more convenient. How to achieve it, I do not know.

**Comment by DexterMagnific** [comment]
> however, during one stage there can be several variables that are not connected with each other and they all need to be checked. In a situation like that reporting all failures at one go would be more convenient

Exactly. Let me explain my case. We a a "component" with many "parameters". We would like to check that when we perform a "set()", it is really taken into account. So we use "get()" to know that.

Our component test case is a series of:

```
err = component.set(param1, refval);
ck_assert_int_eq(err, OK);
val = component.get(param1);
ck_assert_int_eq(refval, val);

component.set(param1, out_of_range_val);
ck_assert_int_eq(err, OUT_OF_RANGE);
```

We call these API tests. They basically check that set/get work and that they don't crash.

We reasonably cannot make a single test case per set/get as it would be very lengthy to write.
Not failing at first assert would be greatly appreciated in such cases.

**Comment by brarcher** [comment]
What you are describing is similar to how the [googletest](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#basic-assertions) c++ framework works, namely there are fatal and non-fatal testing functions. If a non-fatal testing call fails the test continues on (perhaps accumulating failures), and at the end all failing calls are reported. Having used the non-fatal calls in googletest in the past, it was useful to have checks that did not abort the remaining of the test. This way, one test run could find multiple issues (which would be useful if the issues were independent).

Such non-fatal calls could be added to Check. Maybe they could be called "ck_expect_int_eq", etc. Adding support for this would not be simple, as the Check framework currently expects to bail on the first failure, and only supports reporting one issue per test. If someone from the community were interested in investigating this and could propose some changes, I'd be up for reviewing them and giving advice.
